### PR TITLE
Initial changes for BS 1.12.2

### PR DESCRIPTION
**Pull request by ErisApps**
This PR also fixes a NullRef crash related to the HoverHintController, but I'm honestly not sure if it works now as it should so please verify that on your end as well 🙂 

### My last commit adds a wide variety of changes across the board including (but not exhaustive as I might have forgotten some things)

- Zenject for UI. Currently excluding the custom markdown component.
- Fixed a wide variety issues related to the new curved UI.
- Finally fixed the confirmation dialog being so wack.
- Started using BSMT.T and SDK-style csproj (manifest version is now single source of truth and beatmods zips will be created when building in RELEASE mode)
- Added .editorconfig file
- Other changes that I might have missed

Currently relies on a yet-to-be-released BSML version as it fixes something in BSML that would crash the UI.
( https://github.com/monkeymanboy/BeatSaberMarkupLanguage/pull/74 )

Yet, there are still things left to do or fix...
#### Known bugs / TODOs:
- Logger injection in custom components / zenjectifying the custom components
- ~~Wack behavior when trying to disable/enable SingleStartInit mods~~
(Was supposedly fixed in https://github.com/bsmg/BeatSaber-IPA-Reloaded/commit/6b2dc6990a6d75373df648508d26e4b48a8ef468, but feel free to double-check )
- Some markdig components aren't curved correctly just yet
- Scrollbar is gone for one reason or another
- Still need to add confirmation dialog before committing the changeTransaction

### CODE REVIEW

File: ModList/Helpers/Helpers.cs
Comment (nike4613) [inline ModList/Helpers/Helpers.cs]: I mentioned this in DMs, but I'm not a fan of full caps constants. I'd like to keep them as PascalCase.

File: ModList/ModList.csproj
Comment (nike4613) [inline ModList/ModList.csproj:124]: How well is BSMT supported without the extension? I don't use the extension, and prefer my BuildTools solution anyway. That being said, I've been meaning to make it less janky for quite a while now.

File: ModList/Services/ModProviderService.cs
Comment (nike4613) [inline ModList/Services/ModProviderService.cs]: I'm not sure I understand why this needs to be static. Could you explain?

File: ModList/UI/BSML/MarkdownTextHandler.cs
Comment (nike4613) [inline ModList/UI/BSML/MarkdownTextHandler.cs]: These do not need to be exploded like they are. It is so much more readable when each one is on its own line.

File: ModList/UI/BSML/MarkdownTextHandler.cs
Comment (nike4613) [inline ModList/UI/BSML/MarkdownTextHandler.cs:25]: This can use C# 9 and just be `new() ...`

File: ModList/UI/Components/MarkdownText.cs
Comment (nike4613) [inline ModList/UI/Components/MarkdownText.cs]: Is there a reason Helpers needed to be in its own namespace of the same name? Generally I'd do `Utilities.Helpers` to avoid this redundancy, or just have it be in the root namespace if its only the one class.

File: ModList/UI/Components/MarkdownText.cs
Comment (nike4613) [inline ModList/UI/Components/MarkdownText.cs]: I tend to prefer to have the operator at the beginning of the next line rather than the end of the last. Makes it easier to tell at a glance what the condition is.

File: ModList/UI/Components/MarkdownText.cs
Comment (nike4613) [inline ModList/UI/Components/MarkdownText.cs]: I take it the field was made readonly? Do we need to actually set the field most of the time?

Also, `nameof`, please.

File: ModList/UI/Components/TextMeshProUGUILinkBackgroundGenerator.cs
Comment (nike4613) [inline ModList/UI/Components/TextMeshProUGUILinkBackgroundGenerator.cs]: I also mentioned this in DMs; I really dislike the `_` prefix for private fields.

File: ModList/UI/Markdig/RenderHelpers.cs
Comment (nike4613) [inline ModList/UI/Markdig/RenderHelpers.cs]: For stuff like this, its more obvious to me that its an expression switch if the body is indented one more level and the expression body arrow is on the line below the declaration.

File: ModList/UI/Markdig/RenderHelpers.cs
Comment (nike4613) [inline ModList/UI/Markdig/RenderHelpers.cs]: Much the same as above; I prefer the expression bodies, if they aren't trivial, to be on the next line and everything else more indented.

File: ModList/UI/Markdig/UnityRenderer.cs
Comment (nike4613) [inline ModList/UI/Markdig/UnityRenderer.cs]: See previous files.

File: ModList/UI/Markdig/UnityRenderer.cs
Comment (nike4613) [inline ModList/UI/Markdig/UnityRenderer.cs]: The reason I was using this method explicitly rather than as an extension was because I realized that it was more readable not as an extension method. I must have forgotten to make it not an extension method.

File: ModList/UI/Markdig/UnityRenderer.cs
Comment (nike4613) [inline ModList/UI/Markdig/UnityRenderer.cs]: Spaces inside braces please :)

File: ModList/UI/Markdig/UnityRenderer.cs
Comment (nike4613) [inline ModList/UI/Markdig/UnityRenderer.cs]: I typically like having all of the chained method calls aligned with the first, though I don't care about that quite as much as some of the other things I've mentioned.

File: ModList/UI/Markdig/UnityRenderer.cs
Comment (nike4613) [inline ModList/UI/Markdig/UnityRenderer.cs]: If the BCL class is `Uri`, this is `Url`.

File: ModList/UI/Markdig/UnityRendererBuilder.cs
Comment (nike4613) [inline ModList/UI/Markdig/UnityRendererBuilder.cs]: What on earth is this capitalization?

File: ModList/UI/ModListFlowCoordinator.cs
Comment (nike4613) [inline ModList/UI/ModListFlowCoordinator.cs]: Why should the flow coordinator be public?

File: ModList/UI/ModListFlowCoordinator.cs
Comment (nike4613) [inline ModList/UI/ModListFlowCoordinator.cs:90]: It would probably also be worth adding an additional confirmation if they're disabling ModList.

File: ModList/UI/ViewControllers/ModControlsViewController.cs
Comment (nike4613) [inline ModList/UI/ViewControllers/ModControlsViewController.cs]: I generally prefer having the BSML next to the ViewController exactly so that this attribute isn't necessary, as well as generally liking those things to go together since their inextricably tied to each other anyway.

File: ModList/UI/ViewControllers/ModControlsViewController.cs
Comment (nike4613) [inline ModList/UI/ViewControllers/ModControlsViewController.cs]: This should be able to be an expression body method.

File: ModList/UI/ViewControllers/ModalPopupViewController.cs
Comment (nike4613) [inline ModList/UI/ViewControllers/ModalPopupViewController.cs]: This is long enough that I'd want the body on the next line.

**Review by nike4613 — CHANGES_REQUESTED** [review]

File: ModList/ModList.csproj
Comment (ErisApps) [reply ModList/ModList.csproj:124]: It works perfectly fine without the extensions. I'm using Rider so BSMT isn't available for that. 😅

File: ModList/Helpers/Helpers.cs
Comment (ErisApps) [reply ModList/Helpers/Helpers.cs]: I might have missed this when adjusted the .editorconfig file. Will check.

File: ModList/UI/BSML/MarkdownTextHandler.cs
Comment (ErisApps) [reply ModList/UI/BSML/MarkdownTextHandler.cs:25]: C# 9 is sadly not allowed for BeatMods mods for the time being as per request of @Auros

File: ModList/UI/BSML/MarkdownTextHandler.cs
Comment (ErisApps) [reply ModList/UI/BSML/MarkdownTextHandler.cs]: Will look into the .editorconfig file for this.

File: ModList/UI/BSML/MarkdownTextHandler.cs
Comment (nike4613) [reply ModList/UI/BSML/MarkdownTextHandler.cs:25]: That seems... odd. Did he give a reason?

Comment (Auros) [reply ModList/UI/BSML/MarkdownTextHandler.cs:25]: I requested for mods to not use C# 9 since dnSpy can't fully decompile some C# 9 code without it looking funky, and other popular deobfuscators that are used by approvers like JustAssembly will just straight up crash.

If I were the only BeatMods approver i probably wouldn't have an issue and go the extra mile to verify stuff, but I'm not.

Comment (nike4613) [reply ModList/UI/BSML/MarkdownTextHandler.cs:25]: IIRC the only things C# 9 thing that dnSpy *really* doesn't like is function pointers. Most stuff works fine.

File: ModList/UI/ModListFlowCoordinator.cs
Comment (ErisApps) [reply ModList/UI/ModListFlowCoordinator.cs:90]: Can I just add it to the TODO or did I have to actually implement this as well before the PR can be merged?

File: ModList/UI/ModListFlowCoordinator.cs
Comment (ErisApps) [reply ModList/UI/ModListFlowCoordinator.cs]: Accidental change

File: ModList/UI/Markdig/UnityRendererBuilder.cs
Comment (ErisApps) [reply ModList/UI/Markdig/UnityRendererBuilder.cs]: I... erm... might have messed up here...

File: ModList/UI/Components/MarkdownText.cs
Comment (ErisApps) [reply ModList/UI/Components/MarkdownText.cs]: It was indeed marked as readonly, hence the use of  `.SetField<T, U>()`.
That being said, I simply fixed this as it prevented the code from compiling, but I honestly don't know whether this is actually being used at all...

Forgot to use `nameof ()` so will fix that.

File: ModList/Services/ModProviderService.cs
Comment (ErisApps) [reply ModList/Services/ModProviderService.cs]: Explained in DMs. Will need more investigation though.

File: ModList/UI/ModListFlowCoordinator.cs
Comment (nike4613) [reply ModList/UI/ModListFlowCoordinator.cs:90]: As a TODO is probably fine

**Review by nike4613 — APPROVED** [review]

File: ModList/UI/Markdig/UnityRenderer.cs
Comment (ErisApps) [reply ModList/UI/Markdig/UnityRenderer.cs]: .editorconfig has been adjusted for that 😄

**Review by nike4613 — APPROVED** [review]

File: ModList/UI/BSML/MarkdownTextHandler.cs
Comment (ErisApps) [reply ModList/UI/BSML/MarkdownTextHandler.cs:25]: In the meantime, I wrote a approval helper application based on the ILSpy.Decompiler package which can automatically diff between mod versions so C# 9 _should_ be fine to use now if you really feel like it.

**Pull request merged**
